### PR TITLE
Fix depends build on Arch Linux

### PR DESCRIPTION
**Pull request by moetayuko**
## Description

This fixes building tools and dependencies for Android (probably more) on latest Arch Linux

## How Has This Been Tested?

Build passes.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by fuzzard** [comment]
Cheers for fixing that rbpi patch. Everything else looks sane at a glance

File: tools/depends/target/pythonmodule-pil/Makefile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 include ../../Makefile.include
-DEPS= ../../Makefile.include Makefile pillow-crosscompile.patch 01-rbpi-remove-lcms.patch
```
Comment (AlwinEsch) [inline tools/depends/target/pythonmodule-pil/Makefile:2]: If "01-rbpi-remove-lcms.patch" no more needed then it can also be deleted?

**Comment by AlwinEsch** [comment]
The changes are looking OK for me.

Can you also add above in description how exact the build failure was? That we see what was before.

**Comment by dianlujitao** [comment]
Done, patch removed and build error attached to commit message.

**Comment by fuzzard** [comment]
Just a note, The compile error you have put in may potentially be resolved by https://github.com/xbmc/xbmc/pull/18267

An example failure in your log for linking dl, its finding a system library, and not the sysroot usr/lib
```
/home/dianlujitao/Android/Sdk/ndk/21.3.6528147/toolchains/llvm/prebuilt/linux-x86_64/bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: skipping incompatible /usr/lib/libdl.a when searching for -ldl
```

This will be happening due to pkg-config being called from native build pkg-config, and therefore pulling in system pc data. I believe the above PR should correctly tell pkg-config to find the crosscompile pc data for libz and jpeg-turbo, instead of the native/system versions

Feel free to give it a try and let me know if it changes/fixes any behaviour. If you want to link the full log of the pythonmodule-pil build, im guessing you will find that its finding system versions of webp/webpmux/etc in the PIL Setup Summary output

An example of what we want if pkg-config didnt hose us.
```
--------------------------------------------------------------------
PIL SETUP SUMMARY
--------------------------------------------------------------------
version      Pillow 7.1.2
platform     darwin 3.7.4 (default, Aug  6 2020, 11:03:02)
             [Clang 11.0.3 (clang-1103.0.32.62)]
--------------------------------------------------------------------
--- JPEG support available
*** OPENJPEG (JPEG2000) support not available
--- ZLIB (PNG/ZIP) support available
*** LIBIMAGEQUANT support not available
*** LIBTIFF support not available
--- FREETYPE2 support available
*** LITTLECMS2 support not available
*** WEBP support not available
*** WEBPMUX support not available
*** XCB (X protocol) support not available
--------------------------------------------------------------------
```

Thumbs up from me for the removal of the rbpi patch however.

**Comment by dianlujitao** [comment]
I dropped f9c35a37ed45c9e8ea31c22a405a85b0984382cf and picked #18267, but it didn't seem to be changing anything, I got exactly the same linking error as before, where cross compiler seeks for native libs.

PIL summary looks like this:

```
--------------------------------------------------------------------
PIL SETUP SUMMARY
--------------------------------------------------------------------
version      Pillow 7.1.2
platform     linux 3.7.4 (default, Aug 10 2020, 15:17:29)
             [GCC 10.1.0]
--------------------------------------------------------------------
--- JPEG support available
*** OPENJPEG (JPEG2000) support not available
--- ZLIB (PNG/ZIP) support available
*** LIBIMAGEQUANT support not available
*** LIBTIFF support not available
--- FREETYPE2 support available
*** LITTLECMS2 support not available
*** WEBP support not available
*** WEBPMUX support not available
*** XCB (X protocol) support not available
--------------------------------------------------------------------
```

This output is essentially expected and is what I get w/o neither this PR nor #18267, webp and webpmux are marked as unavailable and won't get built.

However just like I mentioned in the commit message, with a 2-step build setting (`build_ext` first and `install-lib` later, separately), `build_ext` will respect our `--disable-foo` settings and print the desired summary above, but `install-lib` will call `build_ext` again w/ default settings, no summary is printed this time. This can be confirmed by directly run the `install-lib` step in the shell ([my output](https://fars.ee/5I92)), even if `--skip-build` is passed in my experiments.

I've no idea why this is happening, maybe it's because we can't pass `--plat-name $(OS)-$(CPU)` to `setup.py install` so it turns to install the default x86 target instead?

**Comment by fuzzard** [comment]
All good then, thanks for testing further

**Review by fuzzard — APPROVED** [review]

**Comment by fuzzard** [comment]
Just looking at cleaning up the makefile a bit. Would you be willing to look at something like the below patch (against current pillow makefile)

removes the now unneeded libdylib variable, consolidates the build/install command into a single instance, rather than the 3 checked against ifeq checks

```
--- Makefile
+++ Makefile
@@ -1,5 +1,5 @@
 include ../../Makefile.include
-DEPS= ../../Makefile.include Makefile pillow-crosscompile.patch 01-rbpi-remove-lcms.patch
+DEPS= ../../Makefile.include Makefile pillow-crosscompile.patch
 
 VERSION.TXT := $(CMAKE_SOURCE_DIR)/version.txt
 APP_NAME=$(shell awk '/APP_NAME/ {print tolower($$2)}' $(VERSION.TXT))
@@ -10,21 +10,17 @@
 SOURCE=$(LIBNAME)-$(VERSION)
 ARCHIVE=$(SOURCE).tar.gz
 
-LIBDYLIB=$(PLATFORM)/dist/Pillow-$(VERSION)-py3.7-$(OS)-$(CPU).egg
 PYTHONPATH=$(PREFIX)/lib/python3.7/site-packages/
+PILPATH=$(PYTHONPATH)
 LDSHARED=$(CC) -shared
 
 ifeq ($(OS),android)
-LIBDYLIB=$(PREFIX)/share/$(APP_NAME)/addons/script.module.pil/lib/PIL/_imaging.so
 PILPATH=$(PREFIX)/share/$(APP_NAME)/addons/script.module.pil
 PYTHONPATH=$(PILPATH):$(PREFIX)/lib/python3.7/site-packages/
 EXTRALDFLAGS=-L$(PREFIX)/lib/dummy-lib$(APP_NAME)/ -l$(APP_NAME) -lm
 else ifeq ($(OS),darwin_embedded)
-LIBDYLIB=$(PREFIX)/share/$(APP_NAME)/addons/script.module.pil/lib/PIL/_imaging.so
 PILPATH=$(PREFIX)/share/$(APP_NAME)/addons/script.module.pil
 PYTHONPATH=$(PILPATH):$(PREFIX)/lib/python3.7/site-packages/
-else ifeq ($(OS),osx)
-LIBDYLIB=$(PLATFORM)/dist/Pillow-$(VERSION)-py3.7-macosx-10.4-x86_64.egg
 endif
 
 ifeq (darwin, $(findstring darwin, $(HOST)))
@@ -32,6 +28,7 @@
 LDSHARED=$(CC) -bundle -undefined dynamic_lookup
 endif
 
+BUILD_OPTS=--plat-name $(OS)-$(CPU) --disable-jpeg2000 --disable-webp --disable-imagequant --disable-tiff --disable-webp --disable-webpmux --disable-xcb --disable-lcms --disable-platform-guessing
 CROSSFLAGS=PYTHONXCPREFIX="$(PREFIX)" CC="$(CC) $(CFLAGS)" LDSHARED="$(LDSHARED)" LDFLAGS="$(LDFLAGS) $(EXTRALDFLAGS)" PYTHONPATH="$(PYTHONPATH)"
 
 all: .installed-$(PLATFORM)
@@ -42,19 +39,18 @@
 $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
 	rm -rf $(PLATFORM)/*; mkdir -p $(PLATFORM)
 	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE)
-ifeq ($(TARGET_PLATFORM),raspberry-pi)
-	cd $(PLATFORM); patch -p1 -i ../01-rbpi-remove-lcms.patch
-endif
 	cd $(PLATFORM); patch -p1 -i ../pillow-crosscompile.patch
 
-$(LIBDYLIB): $(PLATFORM)
-	mkdir -p $(PLATFORM)/output
-	cd $(PLATFORM); $(CROSSFLAGS) $(NATIVEPREFIX)/bin/python3 setup.py build_ext --plat-name $(OS)-$(CPU) --disable-jpeg2000 --disable-webp --disable-imagequant --disable-tiff --disable-webp --disable-webpmux --disable-xcb --disable-lcms --disable-platform-guessing
-
-.installed-$(PLATFORM): $(LIBDYLIB)
+$(PILPATH):
 ifeq ($(OS),android)
 	mkdir -p $(PILPATH)/lib
-	cd $(PLATFORM); $(CROSSFLAGS) $(NATIVEPREFIX)/bin/python3 setup.py install --install-lib  $(PILPATH)
+else ifeq ($(OS),darwin_embedded)
+	mkdir -p $(PILPATH)/lib
+endif
+
+.installed-$(PLATFORM): $(PLATFORM) $(PILPATH)
+	cd $(PLATFORM); $(CROSSFLAGS) $(NATIVEPREFIX)/bin/python3 setup.py build_ext $(BUILD_OPTS) install --install-lib  $(PILPATH)
+ifeq ($(OS),android)
 	cd $(PILPATH)/lib && unzip -o ../Pillow-*.egg
 	cd $(PILPATH)/lib/PIL && \
           sed -i -e 's/import sys/import os, sys /' \
@@ -62,12 +58,8 @@
                  -e 's/pkg_resources.resource_filename(__name__,/os.path.join(os.environ["KODI_ANDROID_LIBS"], /'  _imaging*.py
 	cd $(PILPATH) && rm -rf Pillow-*.egg
 else ifeq ($(OS),darwin_embedded)
-	mkdir -p $(PILPATH)/lib
-	cd $(PLATFORM); $(CROSSFLAGS) $(NATIVEPREFIX)/bin/python3 setup.py install --install-lib  $(PILPATH)
 	cd $(PILPATH)/lib && unzip -o ../Pillow-*.egg
 	cd $(PILPATH) && rm -rf Pillow-*.egg
-else
-	cd $(PLATFORM); $(CROSSFLAGS) $(NATIVEPREFIX)/bin/python3 setup.py install --prefix=$(PREFIX)
 endif
 	touch $@

```

**Comment by dianlujitao** [comment]
The `$(PILPATH)` artifact is different from its real output path, I'm not quite sure but would this mess up the build system? The rest clean-ups LGTM.

**Comment by fuzzard** [comment]
What's the output path for the egg on your system?

**Comment by dianlujitao** [comment]
The `$(PILPATH)` artifact is creating `$(PILPATH)/lib` instead, so in case `$(PILPATH)` exists but `$(PILPATH)/lib` doesn't, the follow-up procedures will fail.

**Comment by fuzzard** [comment]
Thats not an issue, thats behind platform checks (darwinembedded and android only). If you're cross compiling for android the egg is extracted and placed into $(PREFIX)/share/$(APP_NAME)/addons/script.module.pil, which is then copied into the apk by a kodi cmake build script.

For all other platforms, it should just be copying into $(PREFIX)/lib/python3.7/site-packages/ as an egg

$(PREFIX) is the target build location (check tools/depends/Makefile.include for the location)

**Comment by dianlujitao** [comment]
I did test that case, make a full build pass first, `rm -rf ~/android-tools/xbmc-depends/aarch64-linux-android-21-debug/share/kodi/addons/script.module.pil/lib` and make again, it'll fail. I'm unsure if you wanna support such a corner case tho.

**Comment by fuzzard** [comment]
Can you show a complete log of a failure. Jenkins (all PR's, our releases including alpha 1) have used that code without an issue for a month of so.

**Comment by dianlujitao** [comment]
Failure mentioned above:

```
Installed /home/dianlujitao/android-tools/xbmc-depends/aarch64-linux-android-21-debug/share/kodi/addons/script.module.pil/Pillow-7.1.2-py3.7-linux-x86_64.egg
Processing dependencies for Pillow==7.1.2
Finished processing dependencies for Pillow==7.1.2
cd /home/dianlujitao/android-tools/xbmc-depends/aarch64-linux-android-21-debug/share/kodi/addons/script.module.pil/lib && unzip -o ../Pillow-*.egg
/bin/sh: line 0: cd: /home/dianlujitao/android-tools/xbmc-depends/aarch64-linux-android-21-debug/share/kodi/addons/script.module.pil/lib: No such file or directory
make[2]: *** [Makefile:53: .installed-aarch64-linux-android-21-debug] Error 1
make[2]: Leaving directory '/home/dianlujitao/xbmc/tools/depends/target/pythonmodule-pil'
make[1]: *** [Makefile:128: pythonmodule-pil] Error 2
make[1]: Leaving directory '/home/dianlujitao/xbmc/tools/depends/target'
make: *** [Makefile:11: target/.installed-aarch64-linux-android-21-debug] Error 2
```

Note that this only happens with your clean-ups above __and__ manually remove the `lib` folder, a clean full build works fine. This is easy to interpret: `.installed-$(PLATFORM)` depends on `$(PILPATH)` but requires `$(PILPATH)/lib` to exist, make is not smart enough to understand what's been done within the `$(PILPATH)` artifact, once `$(PILPATH)` exists, the requirement is satisfied even if `$(PILPATH)/lib` doesn't, so the follow-up steps will fail.

Fixed by renaming the artifact to `$(PILPATH)/lib`

**Comment by fuzzard** [comment]
ah k, yeah, i see what you're saying. Yeah, i guess adding lib to it should be ok. The worst case (osx/linux for this) would execute that target every build, as the lib folder will not exist. Not sure if that forces its parent rule to trigger every time or not. I'll have a play locally later

**Comment by fuzzard** [comment]
Had a minute to test, and as i was thinking, it was forcing a full rebuild due to $(PILPATH)/lib not existing (osx/linux)

The following gives us the intent, and seems to work locally for me, if you want to give it a try on your end.

https://paste.kodi.tv/edukuhujup.kodi

**Comment by dianlujitao** [comment]
Looks good and compiles fine with android, PR updated.

**Comment by fuzzard** [comment]
Thanks for sticking through this. Greatly appreciated. Looks ok on my end. I'll just call in @wsnipex for a last sanity check whilst jenkins does its thing

**Pull request merged**